### PR TITLE
Grafana integration with all commands

### PR DESCRIPTION
**Pull request by BEAdi**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
fixes: https://github.com/demisto/etc/issues/36960

## Description
Grafana integration.

## Minimum version of Cortex XSOAR
- [ ] 5.5.0
- [ ] 6.0.0
- [ ] 6.1.0
- [x] 6.2.0

## Does it break backward compatibility?
   - [ ] Yes
       - Further details:
   - [x] No

## Must have
- [x] Tests
- [ ] Documentation

### CODE REVIEW

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 1]
```diff
@@ -0,0 +1,771 @@
...
+
+def alerts_list_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    dashboard_id = argToList(args.get('dashboard_id', ''))
+
+    panel_id = args.get('panel_id')
+    query = args.get('query')
+
+    state = argToList(args.get('state', ''))
+    if state:
+        if state not in ALERT_STATES:
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:281]:
```suggestion
    if state and state not in ALERT_STATES:
```

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 2]
```diff
@@ -0,0 +1,771 @@
...
+    return command_results
+
+
+def alert_pause_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    alert_id = str(args.get('alert_id'))
+
+    response = client.alert_pause_request(alert_id, True)
+
+    output = change_key(dict(response), 'alertId', 'id')
+    output.pop('message', None)
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:309]: why?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 3]
```diff
@@ -0,0 +1,771 @@
...
+                                  limit: str = None, page: str = None):
+        params = assign_params(query=query, tag=tag, type=type_, dashboardIds=dashboard_ids,
+                               folderIds=folder_ids, starred=starred, limit=limit, page=page)
+
+        response = self._http_request('GET', 'api/search', params=params, headers=self._headers)
+        response = self._concatenate_url(response)
+
+        return response
+
+    def _concatenate_url(self, response: List[Dict[str, Any]]):
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:169]: Please explain why this is needed

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 4]
```diff
@@ -0,0 +1,771 @@
...
+
+
+''' HELPER FUNCTIONS '''
+
+
+def change_key(response: dict, prev_key: str, new_key: str):
+    """
+    This function changes a key in response, so the values could be aggregated with other functions.
+    """
+    if prev_key in response:
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:188]: Why wouldn't it be?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 5]
```diff
@@ -0,0 +1,771 @@
...
+        response[new_key] = response.pop(prev_key)
+    demisto.debug(f'changing key "{prev_key}" in response to "{new_key}", edited response is: {response}')
+    return response
+
+
+def lower_keys(response: dict):
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:194]:
```suggestion
def keys_to_lowercase(response: dict):
```

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 6]
```diff
@@ -0,0 +1,771 @@
...
+    demisto.debug(f'de-capitalizing key: {s}')
+    return s[0].lower() + s[1:]
+
+
+def url_encode(query: Optional[str]):
+    """
+    Query values with spaces need to be URL encoded e.g. query=Jane%20Doe.
+    """
+    if query:
+        demisto.debug(f'url-encoding query {query}')
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:217]: Are we sure we need this?

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 7]
```diff
@@ -0,0 +1,913 @@
...
+  type: 8
+- display: Trust any certificate (not secure)
+  name: insecure
+  required: false
+  type: 8
+- display: Username
+  name: credentials
+  required: true
+  type: 9
+- additionalinfo: max -> 50
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:23]: Let's rephrase this

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 8]
```diff
@@ -0,0 +1,913 @@
...
+    - default: false
+      description: A comma-separated list of dashboard ids by which to filter the
+        results.
+      isArray: true
+      name: dashboard_id
+      required: false
+      secret: false
+    - default: false
+      description: Panel ids by which to filter the results.
+      isArray: false
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:86]: not an array?

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 9]
```diff
@@ -0,0 +1,913 @@
...
+      name: panel_id
+      required: false
+      secret: false
+    - default: false
+      description: Contained in the name of an alert.
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:91]: what does this mean?

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 10]
```diff
@@ -0,0 +1,913 @@
...
+      secret: false
+    - default: false
+      description: 'A comma-separated list of states by which to filter the results.
+        The options are: all, no_data, paused, alerting, ok, pending.'
+      isArray: true
+      name: state
+      required: false
+      secret: false
+    - default: false
+      description: Limit response to X number of alerts.
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:104]: please look at other limit descriptions from other integrations and copy from them

**Review by guykeller — CHANGES_REQUESTED** [review]

File: Packs/Grafana/IncidentTypes/Grafana.json [hunk 11]
```diff
@@ -0,0 +1,45 @@
+[
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/IncidentTypes/Grafana.json:1]: Done, renamed to `incidenttype-Grafana_Alert.json`

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 12]
```diff
@@ -0,0 +1,771 @@
...
+            'grafana-dashboards-search': dashboards_search_command
+        }
+
+        if command == 'fetch-incidents':
+            first_fetch = params.get('first_fetch', FETCH_DEFAULT_TIME)
+
+            dashboard_id = params.get('dashboard_id', None)
+            panel_id = params.get('panel_id', None)
+            alert_name = params.get('alert_name', None)
+            state = params.get('state', None)
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:748]: Done

File: Packs/Grafana/Classifiers/classifier-Grafana_Alert.json [hunk 13]
```diff
@@ -0,0 +1,20 @@
+{
+    "defaultIncidentType": "Grafana Alert",
+    "description": "",
+    "feed": false,
+    "id": "63460894-c18b-4332-88fd-d5986a9b6904",
```
Comment (barchen1) [inline Packs/Grafana/Classifiers/classifier-Grafana_Alert.json:5]: Please change the id.

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 14]
```diff
@@ -0,0 +1,771 @@
...
+
+HEADERS = {
+    'Accept': 'application/json',
+    'Content-Type': 'application/json'
+}
+
+# In the documentation, the state 'all' is written as 'ALL'. As the latter doesn't work, we use 'all''.
+ALERT_STATES = {'all', 'no_data', 'paused', 'alerting', 'ok', 'pending'}
+
+MAX_INCIDENTS_TO_FETCH = 50
```
[the remark this answers is not in the document]
Comment (barchen1) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:19]: Not sure if we want to fail the test if the input is over 200. maybe add a condition that if the input is above 200 we will set it to 200.
WDYT?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 15]
```diff
@@ -0,0 +1,771 @@
...
+    return incident
+
+
+''' COMMAND FUNCTIONS '''
+
+
+def alerts_list_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    dashboard_id = argToList(args.get('dashboard_id', ''))
+
+    panel_id = args.get('panel_id')
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:276]: Is this arg can also be a list?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 16]
```diff
@@ -0,0 +1,771 @@
...
+    name = args.get('name')
+    login = args.get('login')
+    theme = args.get('theme')
+
+    # login\email is required
+    if not login and not email:
+        raise DemistoException("Login or email must be filled.")
+
+    response = client.user_update_request(user_id, email, name, login, theme)
+    command_results = CommandResults(
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:438]: Maybe we should add a message in the human-readable that the user updated?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 17]
```diff
@@ -0,0 +1,771 @@
...
+    name = args.get('name')
+
+    response = client.teams_search_request(perpage, page, query, name)
+
+    command_results = CommandResults(
+        outputs_prefix='Grafana.Team',
+        outputs_key_field='id',
+        outputs=response['teams'],
+        raw_response=response['teams'],
+        readable_output=tableToMarkdown('Teams', response['teams'], removeNull=True, headerTransform=pascalToSpace)
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:489]:
```suggestion
        readable_output=tableToMarkdown('Teams search results:', response['teams'], removeNull=True, headerTransform=pascalToSpace)
```

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 18]
```diff
@@ -0,0 +1,771 @@
...
+def team_get_by_id_command(client: Client, args: Dict[str, Any]) -> CommandResults:
+    team_id = str(args.get('team_id'))
+
+    response = client.team_get_by_id_request(team_id)
+    command_results = CommandResults(
+        outputs_prefix='Grafana.Team',
+        outputs_key_field='id',
+        outputs=response,
+        raw_response=response,
+        readable_output=tableToMarkdown('Team', response, removeNull=True, headerTransform=pascalToSpace)
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:504]:
```suggestion
        readable_output=tableToMarkdown(f'Team {team_id} results:', response, removeNull=True, headerTransform=pascalToSpace)
```

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 19]
```diff
@@ -0,0 +1,913 @@
...
+      description: Alert eval data
+      type: Unknown
+    - contextPath: Grafana.Alert.executionError
+      description: Alert execution error
+      type: String
+    - contextPath: Grafana.Alert.url
+      description: Alert url
+      type: String
+    - contextPath: Grafana.Alert.evalData.noData
+      description: Alert eval data - has data?
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:170]: Not sure what this field represents.

File: Packs/Grafana/Integrations/Grafana/Grafana.yml [hunk 20]
```diff
@@ -0,0 +1,913 @@
...
+      description: Team email
+      type: String
+    - contextPath: Grafana.User.teams.avatarUrl
+      description: Team avatar url
+      type: String
+    - contextPath: Grafana.User.teams.memberCount
+      description: Team member count
+      type: Number
+    - contextPath: Grafana.User.teams.permission
+      description: Team permission
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.yml:293]:
```suggestion
      description: Number of team permissions.
```

File: Packs/Grafana/Layouts/layoutscontainer-Grafana_Layout.json [hunk 21]
```diff
@@ -0,0 +1,394 @@
...
+            }
+        ]
+    },
+    "group": "incident",
+    "id": "Grafana Layout",
+    "name": "Grafana Layout",
+    "system": false,
+    "version": -1,
+    "fromVersion": "6.0.0",
+    "description": ""
```
Comment (barchen1) [inline Packs/Grafana/Layouts/layoutscontainer-Grafana_Layout.json:393]: Please add a short description

**Review by barchen1 — CHANGES_REQUESTED** [review]

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 17 above)
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:484]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 18 above)
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:499]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 22]
```diff
@@ -0,0 +1,771 @@
...
+
+''' MAIN FUNCTION '''
+
+
+def main():
+    """main function, parses params and runs command functions
+    """
+    params = demisto.params()
+    args = demisto.args()
+    url = params.get('url')
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: It is, changing it. There are no other mandatory args I don't access this way.

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 23]
```diff
@@ -0,0 +1,771 @@
...
+    if last_fetch is None:
+        return first_fetch_dt
+
+    last_fetch = dateparser.parse(last_fetch).replace(tzinfo=utc, microsecond=0)
+    return max(last_fetch, first_fetch_dt)
+
+
+def parse_alerts(alerts: List[Dict[str, Any]], max_fetch: int, last_fetch: datetime):
+    incidents: List[Dict[str, Any]] = []
+    count = 0
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 24]
```diff
@@ -0,0 +1,771 @@
...
+            break
+        # ignoring microsecond because date_to_timestamp doesn't know how to handle it
+        # which causes the last alert to be fetched every time the function is called
+        new_state_date = dateparser.parse(alert['newStateDate']).replace(tzinfo=utc, microsecond=0)
+        incident = parse_alert(alert, new_state_date, last_fetch)
+        if incident:
+            incidents.append(incident)
+            count += 1
+            if new_state_date > updated_last_fetch:
+                updated_last_fetch = new_state_date
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 1 above)
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 25]
```diff
@@ -0,0 +1,771 @@
...
+
+    return updated_last_fetch, incidents
+
+
+def parse_alert(alert: Dict[str, Any], new_state_date: datetime, last_fetch: datetime):
+    if last_fetch:
+        if new_state_date <= last_fetch:
+            return None
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: changed parse_alert as we talked

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 26]
```diff
@@ -0,0 +1,771 @@
...
+        outputs_key_field='id',
+        outputs=response,
+        raw_response=response,
+        readable_output=tableToMarkdown('Dashboard', response, removeNull=True, headerTransform=pascalToSpace)
+    )
+
+    return command_results
+
+
+def test_module(client: Client) -> None:
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: added tests for fetch's parameters

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 27]
```diff
@@ -0,0 +1,771 @@
...
+
+        if command == 'fetch-incidents':
+            first_fetch = params.get('first_fetch', FETCH_DEFAULT_TIME)
+
+            dashboard_id = params.get('dashboard_id', None)
+            panel_id = params.get('panel_id', None)
+            alert_name = params.get('alert_name', None)
+            state = params.get('state', None)
+
+            max_results = arg_to_number(arg=params.get('max_fetch'), arg_name='max_fetch', required=False)
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: 1+2. done
3. then it is set in the fetch function

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 28]
```diff
@@ -0,0 +1,771 @@
...
+            if not max_results or max_results > MAX_INCIDENTS_TO_FETCH:
+                max_results = MAX_INCIDENTS_TO_FETCH
```
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: done

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py]: made the test fail if the value is over the limit

**Review by barchen1 — CHANGES_REQUESTED** [review]

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 29]
```diff
@@ -0,0 +1,799 @@
...
+            message = 'Authorization Error: make sure username and password are correctly set'
+        else:
+            raise e
+    return return_results(message)
+
+
+def fetch_incidents(client: Client, first_fetch: str, dashboard_id: str = None, panel_id: str = None,
+                    alert_name: str = None, state: str = None, limit: Optional[int] = MAX_INCIDENTS_TO_FETCH) -> List[dict]:
+    if not limit:
+        limit = MAX_INCIDENTS_TO_FETCH
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:707]: let's turn this into a one liner

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 30]
```diff
@@ -0,0 +1,799 @@
...
+    last_fetch = demisto.getLastRun().get('last_fetch', None)
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:708]: `None` should not be a specified default value, as it is the default default :)
```suggestion
    last_fetch = demisto.getLastRun().get('last_fetch')
```

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 31]
```diff
@@ -0,0 +1,799 @@
...
+    last_id_fetched = demisto.getLastRun().get('last_id_fetched', 0)
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:709]: Are we sure ID 0 doesn't exist? Are all IDs positive?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 32]
```diff
@@ -0,0 +1,799 @@
...
+        time_start = str(date_to_timestamp(time_start, DATE_FORMAT))
+    time_end = args.get('time_end')
+    if time_end:
+        time_end = str(date_to_timestamp(time_end, DATE_FORMAT))
+
+    tags = argToList(args.get('tags', ''))
+    text = str(args.get('text'))
+
+    response = client.annotation_create_request(text, dashboard_id, panel_id, time_start, time_end, tags)
+    output = dict(response)
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:472]: isn't it already a JSON? what format is the response?

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 33]
```diff
@@ -0,0 +1,799 @@
...
+    dashboard_id = args.get('dashboard_id')
+    panel_id = args.get('panel_id')
+
+    #  epoch numbers in millisecond resolution
+    time_start = args.get('time')
+    if time_start:
+        time_start = str(date_to_timestamp(time_start, DATE_FORMAT))
+    time_end = args.get('time_end')
+    if time_end:
+        time_end = str(date_to_timestamp(time_end, DATE_FORMAT))
```
Comment (guykeller) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:466]: extract to a helper function please

**Review by guykeller — COMMENTED** [review]
some last notes

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 33 above)
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:466]: done, extracted to `set_time_for_annotation`

File: Packs/Grafana/Integrations/Grafana/Grafana.py (hunk 32 above)
Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:477]: changed it to be more readable

**Comment by BEAdi** [comment]
@guykeller can you please review this?

**Review by guykeller — APPROVED** [review]
Consider what we've discussed about tests, other than that looks great!

**Comment by ShirleyDenkberg** [comment]
@BEAdi I reviewed the docs and made several changes to the integration yml file.  Therefore, please regenerate the integration README.
@guykeller @barchen1 Doc review completed.

**Comment by bakatzir** [comment]
what is missing to merge this?

**Review by guykeller — APPROVED** [review]
Great job!

File: Packs/Grafana/Integrations/Grafana/Grafana.py [hunk 34]
```diff
@@ -0,0 +1,842 @@
...
+
+        return response
+
+    def user_add_to_team_request(self, team_id: str, user_id: str):
+        data = {"userId": int(user_id)}
+
+        # 400 - 'User is already added to this team' might be raised - which isn't a real error
+        response = self._http_request('POST', f'api/teams/{team_id}/members', json_data=data, headers=self._headers,
+                                      ok_codes=(200, 400))
+
```
Comment (barchen1) [inline Packs/Grafana/Integrations/Grafana/Grafana.py:120]: Maybe we should think about telling the client that the user is already added to the team when we get `400 code`.

Comment (BEAdi) [reply Packs/Grafana/Integrations/Grafana/Grafana.py:120]: This is what happens - in addition to the `400 code`, Grafana returns a message saying `"Member added to Team"` or `"User is already added to this team"`, that we show to the user.

**Review by barchen1 — APPROVED** [review]

**Pull request merged**